membership-attribute-service/app/actions/AuthAndBackendViaIdapiAction.scala

```diff
@@ -0,0 +1,40 @@
+package actions
+
+import com.gu.identity.{IdapiService, SignedInRecently}
+import components.TouchpointBackends
+import filters.AddGuIdentityHeaders
+import play.api.mvc.{ActionRefiner, Request, Result, Results}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class AuthAndBackendViaIdapiAction(
+  touchpointBackends: TouchpointBackends,
+  howToHandleRecencyOfSignedIn: HowToHandleRecencyOfSignedIn
+)(
+  implicit ex: ExecutionContext
+) extends ActionRefiner[Request, AuthAndBackendRequest] {
+
+  override val executionContext = ex
+  override protected def refine[A](request: Request[A]): Future[Either[Result, AuthAndBackendRequest[A]]] =
+    touchpointBackends.normal.idapiService.RedirectAdvice.getRedirectAdvice(
+      request.headers.get(IdapiService.HeaderNameCookie).getOrElse(""),
+      request.headers.get(IdapiService.HeaderNameIdapiForwardedScope)
+    ).map(redirectAdvice => {
+
+      val backendConf = if (AddGuIdentityHeaders.isTestUser(redirectAdvice.userId)) {
+        touchpointBackends.test
+      } else {
+        touchpointBackends.normal
+      }
+
+      howToHandleRecencyOfSignedIn match {
+        case Return401IfNotSignedInRecently if redirectAdvice.signInStatus != SignedInRecently =>
+          Left(Results.Unauthorized.withHeaders(
+            ("X-GU-IDAPI-Redirect", redirectAdvice.redirect.map(_.url).getOrElse(""))
+          ))
+        case _ =>
```

Comment:
My preference would be to match on each case explicitly. It's clearer why we're taking the given action for each code path and it forces us to think about how we want handle any new ways of "how to handle recency of signed in" that are introduced in the future (especially if we add the compiler option to [fail on warnings](https://stackoverflow.com/questions/28327383/make-compile-fail-on-non-exhaustive-match-in-sbt)).

Reply:
sorry but i don't agree on this point - handling each case explicitly will just lead to repeated code and this is IMO clearer as it is, where its clear there is only one case (comprised of two conditions that must be met) where the response will be a 401 - otherwise the useful details (e.g. identity ID) is passed onto the main handler of the route

Reply:
ok